File: python/remove_duplicates_from_sorted_array_ii.py
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        #read pointer traverses list and to determine if element should be included in final list
        #write pointer adds values that are retained in the correct index
        read_pointer, write_pointer = 0, 0
        #penum and anti_penum track the previous two values to determine whether or not an element is a second duplicate
        penum, anti_penum = None, None
        while read_pointer < len(nums):
            #if the read element is not a duplicate, add it in the appropriate spot and increment the write pointer
            if nums[read_pointer] != anti_penum
                nums[write_pointer] = nums[read_pointer]
                write_pointer += 1
            #update values
            anti_penum, penum, read_pointer = penum, nums[read_pointer], read_pointer + 1
        #remove excess list indices
        for i in range(len(nums) - write_pointer):
            nums.pop()
        return len(nums)

File: java/removeDuplicatesFromSortedArrayII.java
class Solution {
    public int removeDuplicates(int[] nums) {
        //penum and antiPenum store previous values in order to determine whether the present value in the traversal is a duplicate
        int penum = nums[0] - 1;
        int antiPenum = nums[0] - 1;
        //readPointer traverses the array reading each element and determining if it's a duplicate and whether it should be retained
        int readPointer = 0;
        //writePointer points to the indices where the retained elements should appear in the final array
        int writePointer = 0;
        while (readPointer < nums.length) {
            //if the read value isn't a duplicate, insert it into the array at the correct spot
            if (nums[readPointer] != antiPenum) {
                nums[writePointer] = nums[readPointer];
                writePointer++;
            }
            //update the other values
            antiPenum = penum;
            penum = nums[readPointer];
            readPointer++;
        }
        return writePointer;
    }
}